elastic/beats
# Remove nil-zero metrics and linux-exclusive metrics from Metricbeat

## What does this PR do? 

This PR has two main parts:

- If a metric is currently reporting a "null zero" (that is, the metric doesn't exist for an OS, but we report a `0` instead of nothing) for a given metric on a given OS, add some conditional logic so we don't report it.
- If we're reporting a group of metrics that is highly OS-specific, _add that metric group to the linux module, and don't report it if we're running under fleet_. This is because we want to move some of these metrics, but don't want to break metricbeat. As Fleet is beta, we want to introduce some needed changes while we can.

Also note, I'm actively working on this PR, so it might be a tad rough right now, but I want eyes on it.

## Where do we go from here?

- A large part of this PR is essentially a holdover until we can make breaking changes to reported fields in metricbeat. Once we hit 8.0, we can remove the `IsAgent` logic we have to prevent breaking field changes in metricbeat. 
- Once this is merged, the additional linux metricsets will be added to the Fleet integration. 

## Open Questions

- We don't want to pull metrics out from under user's feet. To accomplish this, any linux-exclusive metrics we're reporting in `system` now have their own module in `linux`. After this PR, we'll add these new linux modules to the Linux integration within fleet. From the perspective of metricbeat, we're now reporting identical data in two places: in the default system module, and in new metricsets within the linux integration. Is this the best route? Is there a better way to handle metrics we're trying to relocate?

## Why is it important?

This is the result of a lot of work between @mukeshelastic and I, with two goals:

- Create a leaner system Integration for fleet that contains only cross-compatible datasets
- Clean up the metricsets so we don't report "null zeros" on operating systems where a given metric doesn't exist. 

## Checklist


## Author's Checklist

- [x] Find a way to run this under Agent
- [x] Test with the rest of the elastic stack
- [x] Test under Windows

## How to test this PR locally

- Pull PR and build via `mage package` to generate a system-specific build object. You can use the `PLATFORMS` environment variable to constrain to a given OS/arch pair like `linux/amd64`.
- Running as metricbeat (no Fleet stack), test a few things depending on your OS:
  - Linux:
    - Make sure the `cpu`, `memory`, `diskio`, `process`, `filesystem`, `fsstat` fields _haven't changed_ relative to `data.json`
    - Test the new `memory` and `iostat` linux metricsets for all the expected fields.
  - Windows:
    - Using the spreadsheet linked above, check the results of the  `cpu`, `memory`, `diskio`, `process`, `filesystem`, and`fsstat`  metricsets to make sure all the fields reported as `Incorrect 0 value?` on the spreadsheet are no longer reported.
- If you can somehow connect your custom build to Agent/Fleet, check to make sure the `IsAgent` constraints are working and we're _not_ reporting additional data.
